### ISSUE TITLE
CREATE ASSEMPLY script leads to  "SQL70557: This assembly is corrupt or not valid" on dotnet build

### ISSUE THREAD

**Issue by lausitzer** [microsoft/DacFx#523]
- SqlPackage or DacFx Version: 162.4.92.3
- .NET Framework (Windows-only) or .NET Core:  .NET 8.0
- Environment (local platform and source/target platforms): SQL sdk project (sdk version 0.2.4-preview) /tp:Sql150 on Windows 11

**Steps to Reproduce:**

1. Create new SQL project for SQL-Server 2019:  dotnet new sqlproj -tp Sql150
2. Write SQL-Script for CREATE ASSEMBLY containing your assembly as binary literal, e.g.: CREATE ASSEMBLY [TestCLR] AUTHORIZATION [dbo] FROM 0x4D5A90000300000...
3. Run:  dotnet build

Build fails with "Build error SQL70557: This assembly is corrupt or not valid" pointing to your CREATE ASSEMBLY script.
For a cross check, executed CREATE ASSEMBLY script with SSME. Assembly gets imported flawless if appropriate conditions are met (CLR-support enabled, login for assemblies signing key created).

Also tried copying dll to $(IntermediateOutputPath) and $(OutputPath) and using assemblies name instead off binary literal. Leads to SQL70502: The assembly source is not valid. Only binary literals are allowed .

Did not test on prior versions of SDK.

Find sample-project attached. [MySqlProj.zip](https://github.com/user-attachments/files/17744602/MySqlProj.zip)

(DacFx/SqlPackage/SSMS/Azure Data Studio)

**Comment by lausitzer** [comment]
Maybe related: https://github.com/microsoft/azuredatastudio/issues/21370

**Comment by dzsquared** [comment]
Unfortunately, SQLCLR is based only on .NET Framework, we can't support it on SQL projects with .NET 8 unless significant changes are done to the SQLCLR. To build a SQL project (including SDK-style SQL projects) with SQLCLR assemblies you would need to build it with msbuild and/or Visual Studio.

![Image](https://github.com/user-attachments/assets/cbb300fe-9ed9-468a-a0db-0c969fc783c7)



![Image](https://github.com/user-attachments/assets/be2b47c8-df07-4f66-bcc0-49507d205c18)

This same project fails to build with dotnet build (.NET 8 SDK).
(noting that SDK 0.2.4-preview has a bug with SSDT, so I downgraded to 0.2.3-preview for testing with Visual Sutdio)

In a pipeline, you'd use an msbuild task like - https://learn.microsoft.com/en-us/azure/devops/pipelines/tasks/reference/msbuild-v1?view=azure-pipelines - to build a SQL project.


... unfortunately, this does mean you are restricted to Windows build agents only.



I'd encourage you to share feedback on SQLCLR to the SQL engine team via https://aka.ms/sqlfeedback



ps - thank you thank you for the sample project attachment! love it when digging into an issue is made simpler with the repro setup.

**Comment by lausitzer** [comment]
@dzsquared Many thanks for investigation. To be clear, its not intended to build the CLR-assembly using dotnet build. I just want to integrate an external build assembly into my dacpac. From my perspective there is no need to parse this assembly with .NET 8. However, I don't know what's going on inside that causes this behavior. Wouldn't it be possible to accept the assembly as is?

**Comment by tn-5** [comment]
@lausitzer it would be really nice if it were possible to include a pre-build assembly in the "dotnet build" pipeline. Currently we work around in a rather ugly way by defining normal SQL functions for those defined in the CLR assembly and then in the post deploy we drop  those, create the assembly and recreate all the functions. At least it works but I agree that it would be much better if one could just integrate a pre-build CLR dll into the dotnet build process.

**Comment by lausitzer** [comment]
Suggestion to the SQL database team: https://feedback.azure.com/d365community/idea/e90dcf7a-f5ac-ef11-95f6-000d3a059eeb

**Comment by PaulGarnerUk** [comment]
Is it possible someone could document the workaround?  I was encouraged by the documentation to start a new SQL Project using the new SDK style dotnet build technique, but it turns out the (existing) database I'm bringing into the project has a single CREATE ASSEMBLY function.

Because SQL CLR isn't supported (yet) it means I can't include the assembly. Which then means the CREATE FUNCTION that relies on the assembly also can't be included. Which then means the numerous stored procedures that rely on the function can't be included etc etc.  It makes the whole point of bringing the database into a build process pointless.

If there was a way to mark the CREATE FUNCTION with a 'trust me' attribute that didn't attempt to validate the existence of the Assembly whilst building - that would be fine (and I could then publish the dacpac and not remove existing assemblies).   But I can't see a way to achieve this.

Long story short - not supporting SQL CLR and asking people to start using net core based projects instead seems like it's waay too early unless there are documented workarounds for these type of issues.

**Comment by dzsquared** [comment]
@PaulGarnerUk - thank you for pinging on this. started a small docs update with this topic, clarifying the .NET Framework build requirement.

I do want to explore the project architecture where the assembly is built by itself in a project with msbuild (.net framework) and that .dacpac is used as a database or package reference in other projects. I haven't had a chance to explore this fully to see what flexibility it creates, but wanted to be candid in how I'm working on flexibility in spite of the SQL engine dependency on .NET Framework for SQL CLR.

**Comment by ErikEJ** [comment]
> asking people to start using net core based projects and not supporting SQL CLR seems like a heck of a gamble

A lot of people have carefully stayed away from SQL CLR so "a heck of a gamble" is maybe a slight exaggeration.

**Comment by dzsquared** [comment]
I absolutely accept the constructive criticism that a tradeoff like SQLCLR being .NET Framework only should have the workaround(s) documented. This can be so much better.

Yeah, SQLCLR is a little different, but its a part of SQL Server so we're not just ripping it out of SQL projects. I can't promise it'll be mainstream in SQL projects though, because as @ErikEJ mentioned SQLCLR is used quite a bit less than most other features.

**Comment by ErikEJ** [comment]
I am working on a blog post and sample 

@jmezach Hopefully we can then close https://github.com/rr-wfm/MSBuild.Sdk.SqlProj/issues/109

**Comment by PaulGarnerUk** [comment]
Thanks. I appreciate the feedback and ideas. I'm very green to this area and likely naïve, but my thoughts were similar to an earlier comment.  I'm not intending to build the assembly from code in this project. I simply want a way to create it (and potentially update it).  That means I'm perfectly happy with creating the binary from raw hex values, and so I assume the problem with this is that somehow the dotnet build process is trying to validate the assembly that is created (and discovering it requires full fat framework)?

Ideally there'd be a way to bypass this validation.  I'm very much under the impression that SQL CLR will not be updated any time soon, so is it feasible that if the dotnet build encounters a CREATE ASSEMBLY with raw hex values that it just does not attempt to validate it?


That 'heck of a gamble' statement was not intended to antagonise. Apologies if it came across the wrong way. What I was trying to explain is that if we (or anyone) adopted the new dotnet build sdk approach (as encouraged by documentation). And then months later found we really need to add a SQL CLR function (such as in our case, to add regex support to queries) it would be incredibly frustrating to realise the issues at that point.  

For the moment I have our CI pipeline configured as a Windows agent and running a VSBuild@1 task, which seems happy enough. I'm just left in an awkward position where the development experience is hacky.  The advice to experiment with building the assembly in a separate dacpac was a great idea and I may investigate that as time allows.

**Comment by ErikEJ** [comment]
@PaulGarnerUk My blogpost will show a way to seperate the SQL CLR object build from your SDK based build, so you can seperate and only need to build on Windows if there are updates to the SQL CLR objects.

**Comment by ErikEJ** [comment]
Blog post (and sample) published now - looking forward to feedback!

https://erikej.github.io/dacfx/sqlclr/2025/01/28/dacfx-sqlclr-msbuild-sdk-sqlproj.html

**Comment by ErikEJ** [comment]
@lausitzer @PaulGarnerUk Would be interested to hear if my "wrokaround" is useful

**Comment by PaulGarnerUk** [comment]
Many thanks. It's a great idea. I understood the concept as soon as @dzsquared mentioned it, and the additional documentation about the process is very welcome.  Now all I need is a gap in my schedule to experiment with it more!

**Comment by lausitzer** [comment]
@ErikEJ Thanks for your blog post. I tried this in an slightly differnt manner using ArtifactReferences instead of PackageReferences and reworked my solution. I packed my SQLCLR-assemblies in theire own solution in minimal msbuild based projects containing just them and the stuff needed to get them imported in a fresh sql-container (create asymmetric key and login from signing key etc.). Now they get build in a separat first build step on a VSBuild-container and publish theire artifacts for the main build. All the other stuff was migrated to sdk-style projects. By testing in my development environment I can confirm that your solution works fine (tested until deployment dacpacs to docker). To complete the transformation I'm strugling now with https://github.com/microsoft/DacFx/issues/556.

**Comment by gfody** [comment]
> Blog post (and sample) published now - looking forward to feedback!
> 
> https://erikej.github.io/dacfx/sqlclr/2025/01/28/dacfx-sqlclr-msbuild-sdk-sqlproj.html

I tried this w/Microsoft.Build.Sql 1.0.0-rc1 and the first problem was that after installing SSDT preview the old projects no longer load without upgrading them. They crash with:
```
The 'Microsoft SQL Server Data Tools' package dit not load correctly.

SetSite failed for package [Microsoft SQL Server Data Tools]Source: 'mscorlib' Description: An item with the same key has already been added.
System.ArgumentException: An item with the same key has already been added.
   at System.ThrowHelper.ThrowArgumentException(ExceptionResource resource)
   at System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add)
   at Microsoft.VisualStudio.Data.Tools.Package.DesignerServices.DatabaseChangesManager.AddMasterReferenceFile(SqlPlatforms platform)
   at Microsoft.VisualStudio.Data.Tools.Package.DesignerServices.DatabaseChangesManager.Initialize()
   at Microsoft.VisualStudio.Data.Tools.Package.UI.DataPackage.<InitializeAsync>d__88.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.Shell.AsyncPackage.<>c__DisplayClass17_0.<<Microsoft-VisualStudio-Shell-Interop-IAsyncLoadablePackageInitialize-Initialize>b__1>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.VisualStudio.Services.VsTask.RethrowException(AggregateException e)
   at Microsoft.VisualStudio.Services.VsTask.InternalGetResult(Boolean ignoreUIThreadCheck)
```

..I couldn't get both to work at once so I just converted the old db projects to class library projects targeting net472 and added them to the new database projects as project references. I manually added all of the "as external name" functions and sprocs to the new database project and I feel like this is something I'm willing to compromise on (it's nice to be able to specify default values and specific datatypes that the SqlFunctionAttribute doesn't support, and it's nice to see our CLR triggers inline with their table definitions rather than appearing seemingly magically out of nowhere)

So the class library to become CLR assembly looks like this:
```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <AssemblyName>MyCLR</AssemblyName>
    <TargetFramework>net472</TargetFramework>
    <DisableImplicitFrameworkReferences>true</DisableImplicitFrameworkReferences>
  </PropertyGroup>
  <ItemGroup>
    <Reference Include="System" />
    <Reference Include="System.Core" />
    <Reference Include="System.Data" />
    <Reference Include="System.Xml" />
  </ItemGroup>
</Project>
```

And in the database project I added a target to generate the "create assembly" statement for build (I had to exclude `$(AssemblyName).dll` because an empty/invalid one shows up in the output path:
```xml
  <ItemGroup>
    <ProjectReference Include="..\MyCLR\MyCLR.csproj" />
  </ItemGroup>

  <Target Name="Assemblies" AfterTargets="CoreCompile">
    <ItemGroup>
      <ReferencedAssemblies Include="$(OutputPath)*.dll" Exclude="$(OutputPath)$(AssemblyName).dll" />
    </ItemGroup>
    <PropertyGroup>
      <CreateAssemblyScript>$(IntermediateOutputPath)CreateAssemblies.sql</CreateAssemblyScript>
      <PwshCommand>
        $hex = [Convert]::ToHexString([IO.File]::ReadAllBytes($_.FullName))
        $sql = "create assembly [%24($_.BaseName)]`nfrom 0x$hex`nwith permission_set = unsafe`ngo`n"
        [IO.File]::AppendAllText("$(CreateAssemblyScript)", $sql)
      </PwshCommand>
    </PropertyGroup>
    <Exec Command="pwsh -c 'Get-Item %(ReferencedAssemblies.Identity) | % { $(PwshCommand) }'" IgnoreExitCode="true" WorkingDirectory="$(MSBuildProjectDirectory)" />
    <ItemGroup>
      <Build Include="$(CreateAssemblyScript)" />
    </ItemGroup>
  </Target>
```

..this all seems to work fine if I do clean builds from within VS, but `dotnet build` fails with this error:
```
MyCLR.csproj targets 'net472'. It cannot be referenced by a project that targets '.NETFramework,Version=v4.6'.
```
I've tried adding TargetFramework to the database project but doesn't help. Occasionally the build from VS will fail with this error:
```
Imports contains an invalid framework: 'unsupported' in 'project.json'.
```
and persists until I delete `obj/project.assets.json`

I tried `1.0.0-rc2` but that fails immediately with this error message:
```
.nuget\packages\microsoft.build.sql\1.0.0-rc2\Sdk\Sdk.targets(79,5): error MSB4036: The "CheckForDuplicateItems" task was not found. Check the following:
 1.) The name of the task in the project file is the same as the name of the task class.
 2.) The task class is "public" and implements the Microsoft.Build.Framework.ITask interface.
 3.) The task is correctly declared with <UsingTask> in the project file, or in the *.tasks files located in the "C:\Program Files\Microsoft Visual Studio\2022\Professional\MSBuild\Current\Bin\amd64" directory.
```

**Comment by gfody** [comment]
> [@gfody](https://github.com/gfody) Looks like you did not read the blog post properly. 😄
> 
> And yes, you must have a seperate VS install that only uses the old project style, you cannot have both.
> 
> (The VS Community license allows you to use old Database Projects in any licensing scenario btw.)
> 
> According to my blog post, use an legacy / classic .sqlproj for the project with the SQLCLR objects,

I tried to follow it in spirit but using Microsoft.Build.Sql rather than MSBuild.SDK.SqlProj since you said the same approach might work. After encountering the problem of not being able to use new and old sqlproj at the same time I had to deviate. It didn't occur to me to try installing another edition of VS but that isn't something I'd want to instruct my team to do. FWIW I would prefer to be able to build everything with one edition of VS and am willing to sacrifice the functionality of the old project style.

> it will never work with the SDK project style, I do not know what gave you that idea - let me know if the blog post is unclear in that respect.

The SDK style CLR assembly does work, ie adding the binary to the database project as "create assembly" statement satisfies the assembly references and the database project builds and produces a dacpac with the assembly included. What isn't working is "dotnet build" but the error messages seem like nonsense and not anything that fundamentally shouldn't work - but maybe I'm missing something cause you say it will never work, but what do you mean by "it"?

**Comment by ErikEJ** [comment]
@gfody You are deviating too much from my proposed  and working solution...

**Comment by gfody** [comment]
> [@gfody](https://github.com/gfody) You are deviating too much from my proposed and working solution...

okay I gave this some more time and less deviation and can confirm the workaround works for both MSBuild.SDK.SqlProj and Microsoft.Build.Sql. I was able to get an all-in-one dotnet build by making a [janky dacpac target](https://gist.github.com/gfody/01d584e1dfef73e572023a551f8fd480) maybe that can help w/testing.

**Comment by ErikEJ** [comment]
@gfody Thansk a lot for confirming!

**Comment by SaxoLigtorn** [comment]
For my SQLCLR solution, I managed to build Microsoft.Build.Sql based projects using **_msbuild_** instead of dotnet build. MSBuild seems to be .netframework based, and that seems control, which target files get used from the nuget Microsoft.Build.Sql

One caveat is that my projects, needs to target net472. Also I am not sure if it can build on linux or not. I did manage to build in Visual Studio 2026, with all it shortcomings

The strategy is to use a standard sdk styled csproj for building the clr assembly and generate a sql file, which contains the assembly as a hexstring, which the main sqlproj uses. 

My clr project build and runs a powershell command afterwards, that converts the generated assembly into the SQL script

```xml
<Project Sdk="Microsoft.NET.Sdk">
	<PropertyGroup>
		<AssemblyName>Soendergaard.DataMasking.CLR</AssemblyName>
		<GeneratedClrAssemblySqlPath>$(MSBuildProjectDirectory)\..\Soendergaard.DataMasking.Database.DataMasking\ClrModule\Assemblies\Soendergaard.DataMasking.CLR.sql</GeneratedClrAssemblySqlPath>
		<SignAssembly>true</SignAssembly>
	</PropertyGroup>

	<Target Name="GenerateClrAssemblySqlForSqlProject" AfterTargets="Build">
		<Exec Command='pwsh.exe -NoProfile -NonInteractive -ExecutionPolicy Bypass -File "$(MSBuildProjectDirectory)\Generate-ClrAssemblySql.ps1" -ClrAssemblyPath "$(TargetPath)" -OutputSqlPath "$(GeneratedClrAssemblySqlPath)" -AssemblyName "$(AssemblyName)" -PermissionSet "UNSAFE"' />
	</Target>
</Project>
``` 

This is the powershell scripts it executes

```powershell
param
(
    [Parameter(Mandatory = $true)]
    [string]$ClrAssemblyPath,

    [Parameter(Mandatory = $true)]
    [string]$OutputSqlPath,

    [string]$AssemblyName = "Soendergaard.DataMasking.CLR",

    [ValidateSet("SAFE", "EXTERNAL_ACCESS", "UNSAFE")]
    [string]$PermissionSet = "UNSAFE"
)

$ErrorActionPreference = "Stop"

Write-Host "###################################################################"
Write-Host "##### Generate-ClrAssemblySql.ps1 parameters:"
Write-Host "###################################################################"
Write-Host "  ClrAssemblyPath : $ClrAssemblyPath"
Write-Host "  OutputSqlPath   : $OutputSqlPath"
Write-Host "  AssemblyName    : $AssemblyName"
Write-Host "  PermissionSet   : $PermissionSet"

if (-not (Test-Path -LiteralPath $ClrAssemblyPath))
	{ throw "CLR assembly not found: $ClrAssemblyPath"}

$bytes = [System.IO.File]::ReadAllBytes($ClrAssemblyPath)
$hexBuilder = New-Object System.Text.StringBuilder
[void]$hexBuilder.Append("0x")
foreach ($b in $bytes)
	{ [void]$hexBuilder.Append($b.ToString("X2"))}

$targetDir = Split-Path -Parent $OutputSqlPath
if (-not [string]::IsNullOrWhiteSpace($targetDir))
	{ New-Item -Path $targetDir -ItemType Directory -Force | Out-Null}

$sql = @"
CREATE ASSEMBLY [$AssemblyName] AUTHORIZATION [dbo]
FROM $($hexBuilder.ToString())
WITH PERMISSION_SET = $PermissionSet;
"@

Set-Content -Path $OutputSqlPath -Value $sql -Encoding UTF8 -Force
Write-Host "Generated assembly SQL: $OutputSqlPath"
``` 

The real sqlproj which uses the CLR assembly looks like this

```xml
<Project DefaultTargets="Build">
  <Sdk Name="Microsoft.Build.Sql" Version="2.1.0" />
  <PropertyGroup>
    <DSP>Microsoft.Data.Tools.Schema.Sql.Sql170DatabaseSchemaProvider</DSP>
    <Name>Soendergaard.DataMasking.Database.DataMasking</Name>
  </PropertyGroup>
  <ItemGroup>
    <Build Remove="Scripts\*.sql" />
    <None Include="Scripts\Script.PreDeployment.AsymmetricKeySetup.sql" CopyToOutputDirectory="PreserveNewest" />
    <None Include="Templates\AssemblySecurityTemplate.sql" />
    <PostDeploy Include="Scripts\Script.PostDeployment.sql" />
    <PreDeploy Include="Scripts\Script.PreDeployment.sql" />
  </ItemGroup>
</Project>
``` 

this contains sql scripts for normal sql objects, but also manual script files for the mapping to assembly functions like these. In the old project, those were automatically handled, but with this solution, you have to manage them yourself

```tsql
CREATE PROCEDURE [ClrModule].[CreateDirectory]
@Path NVARCHAR (MAX) NULL, @ShowOutput BIT NULL, @SimulateOnly BIT NULL
AS EXTERNAL NAME [Soendergaard.DataMasking.CLR].[Soendergaard.DataMasking.CLR.ClrClasses.ClrFileFunctions].[CreateDirectory]
```
